### PR TITLE
Introduce a random sleep in the Netdata updater

### PR DESCRIPTION
**Pull request by prologic**
##### Summary

ssia

##### Component Name

- area/packaging

##### Test Plan

Tested the shell expression:

```#!sh
sleep $(((RANDOM % 10) + 1))s
```

##### Additional Information

This will hopefully aliviate the stampede like effect we're seeing in Netdata
Cloud Production where we see spikes of disconnections and reconenctions.
The timing lines up with the default Debian crontab entry for "daily" jobs
and thus adding a random sleep into the updater should help spread this out
a bit and reduce load on our system9s) in short intervals.

### CODE REVIEW

**Comment by cakrit** [comment]
The sleep command is too early in the script, there are a couple of possible failure conditions that would prevent installation further down. It should probably go before the actual running of the installation code, after the checksum validation. 

This is a bad hack, I don't like it at all. Have we really exhausted all other options and have to resort to this?

**Comment by Ferroin** [comment]
> This is a bad hack, I don't like it at all. Have we really exhausted all other options and have to resort to this?

There is _zero_ functionality to provide this in any of the task schedulers we do or could use for our updater. The expectation is that if a job needs random jitter like this, it will handle it itself.. I'm not hugely fond of it myself (it should ideally be set up so it only runs when run non-interactively, and I'd go with a much longer range (more like 5-10 minutes, at least)), but it's better than nothing.

Aside from the stampede effect we're seeing in Netdata Cloud due to updates, it's also good practice to help avoid putting undue load on our update infrastructure.

Also, this is generally expected behavior for automated update systems. See for example FreeBSD's update tool (it refuses to run non-interactively without special handling, and the cron job mode sleeps a random number of seconds up to a full hour before actually doing it's job).

**Comment by cakrit** [comment]
Ok, fine, just move it to after the sanity checks.

**Comment by cakrit** [comment]
Fine, go ahead. I see we are trying to spread out the downloads themselves as well, so the sleep can't go much later than where it is...

File: packaging/installer/netdata-updater.sh [hunk 1]
```diff
@@ -210,6 +210,10 @@ tmpdir=
 
 trap cleanup EXIT
 
+# Random sleep to aileviate stampede effect of Agents upgrading
+# and disconnecting/reconnecting at the same time (or near to).
+sleep $(((RANDOM % 10) + 1))s
+
```
Comment (Ferroin) [inline packaging/installer/netdata-updater.sh:216]: On further thought, I'm very strongly of the opinion that this should:

1. Be used only when run non-interactively (no controlling terminal for the script is sufficient IMO).
2. Be _much_ longer on the maximum, probably at least 5 minutes, but more ideally closer to half an hour or an hour.

Comment (ilyam8) [reply packaging/installer/netdata-updater.sh:216]: probably we need something like that

```
min = x
max = y
sleep = rand(max - min) + min
```

Comment (Ferroin) [reply packaging/installer/netdata-updater.sh:216]: Provided it's a uniform distribution (which I believe is the case for the bash implementation of `$RANDOM`), we should just set the minimum at 0.

Comment (prologic) [reply packaging/installer/netdata-updater.sh:216]: Unclear of the unformity of Bash's `$RANDOM` but the man page has this to say:

>        RANDOM Each time this parameter is referenced, a random integer between 0 and 32767 is generated.  The sequence of random numbers may  be
>              initialized  by  assigning  a  value  to  RANDOM.  If RANDOM is unset, it loses its special properties, even if it is subsequently
>              reset.

Comment (mfundul) [reply packaging/installer/netdata-updater.sh:216]: Wait, so unsetting `$RANDOM` is a security concern? Like a shell could export `$RANDOM` to a subshell and kill the randomness?

Comment (Ferroin) [reply packaging/installer/netdata-updater.sh:216]: @mfundul Yes. Most 'special' shell variables are only special until they're overridden by the environment. It's kind of like how some old languages let you use purely numeric identifiers.

Comment (prologic) [reply packaging/installer/netdata-updater.sh]: Done

File: packaging/installer/netdata-updater.sh [hunk 2]
```diff
@@ -210,6 +210,13 @@ tmpdir=
 
 trap cleanup EXIT
 
+# Random sleep to aileviate stampede effect of Agents upgrading
+# and disconnecting/reconnecting at the same time (or near to).
+# But only we're not a controlling terminal (tty)
+if [ ! -t 1 ]; then
+  sleep $((RANDOM % (3600 - 1800 + 1) + 1800))s
```
Comment (Ferroin) [inline packaging/installer/netdata-updater.sh:217]: Unless I'm mistaken, this results in a range of 30-60 minutes for the delay. Is there some reason not to just make it 0-60 minutes instead if we're putting 60 minutes as the upper bound?

Comment (prologic) [reply packaging/installer/netdata-updater.sh:217]: Its called "middle ground" I met you halfway. 😀

Comment (ilyam8) [reply packaging/installer/netdata-updater.sh:217]: > Unless I'm mistaken, this results in a range of 30-60 minutes for the delay.

this is not clear from the code, i sugges to add it as a code ([min, max])

Comment (Ferroin) [reply packaging/installer/netdata-updater.sh:217]: @prologic I think I might have been unclear then in what I was saying. I still think the lower limit should be 0 seconds, I just believe that we should spread across a much longer period of time than the 10s that your original code did.

If we really are just going to do a 30 minute range, then that should be 0-30 minutes, not 30-60.

**Comment by prologic** [comment]
Sanity check before merging...

```#!sh
root@ubuntu2004:~# tail -f /var/log/syslog
May 26 00:45:01 ubuntu2004 CRON[81223]: (root) CMD (/usr/libexec/netdata/netdata-updater.sh)
root@ubuntu2004:~# ps aux | grep netdata-updater
root       81223  0.0  0.0   2608   544 ?        Ss   00:45   0:00 /bin/sh -c /usr/libexec/netdata/netdata-updater.sh
root       81224  0.0  0.3   7024  3412 ?        S    00:45   0:00 bash /usr/libexec/netdata/netdata-updater.sh
root       81227  0.0  0.0   6436   740 pts/0    S+   00:45   0:00 grep --color=auto netdata-updater
root@ubuntu2004:~# strace -p 81224
strace: Process 81224 attached
wait4(-1, ^Cstrace: Process 81224 detached
 <detached ...>
root@ubuntu2004:~# ps aux | grep sleep
root       81225  0.0  0.0   5476   592 ?        S    00:45   0:00 sleep 336s
root       81234  0.0  0.0   6432   724 pts/0    S+   00:46   0:00 grep --color=auto sleep
root@ubuntu2004:~# pstree -p 81223
sh(81223)───bash(81224)───sleep(81225)
```

Behaving as intended via CronD.

File: packaging/installer/netdata-updater.sh [hunk 3]
```diff
@@ -185,8 +185,8 @@ update() {
       do_not_start="--dont-start-it"
     fi
 
-    if [ -n "${NETDATA_SELECTED_DASHBOARD}" ] ; then
-        env="NETDATA_SELECTED_DASHBOARD=${NETDATA_SELECTED_DASHBOARD}"
+    if [ -n "${NETDATA_SELECTED_DASHBOARD}" ]; then
+      env="NETDATA_SELECTED_DASHBOARD=${NETDATA_SELECTED_DASHBOARD}"
     fi
```
Comment (Ferroin) [inline packaging/installer/netdata-updater.sh:190]: Not critical for this PR, but this block can actually be removed. It's left over from when we had the switcher script for the dashboard, I just forgot to remove it when that got removed.

Comment (prologic) [reply packaging/installer/netdata-updater.sh:190]: Put up a PR to clean this up then :)

**Comment by prologic** [comment]
Travis CI failures are becuase of:

```
00:00:49.878291  Running BATS file..
00:00:49.896820  1..3
00:02:22.894917  ok 1 install stable netdata using kickstart
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received
The build has been terminated
```

Unintended side-effect of introducing a random delay.

**Comment by prologic** [comment]
This can be fixed by disabling Bash's `$RANDOM` special variable by doing:

```#!sh
unset RANDOM; export RANODM=0
```

This permeates to subshells too;

```#!sh
$ unset RANDOM; export RANDOM=0; ( echo $RANDOM ); /bin/sh -c 'echo $RANDOM'
0
14598
```

**Pull request merged**